xds/src/main/java/com/linecorp/armeria/xds/ResourceHolderWithPrimer.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.xds;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.envoyproxy.envoy.config.cluster.v3.Cluster;
+import io.envoyproxy.envoy.config.endpoint.v3.ClusterLoadAssignment;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.envoyproxy.envoy.config.route.v3.RouteConfiguration;
+
+/**
+ * A resource holder that has its primer.
+ *
+ * @param <SELF> the type of this
+ * @param <T> the type of the resource. Can be {@link Listener}, {@link RouteConfiguration}, {@link Cluster} or
+ *            {@link ClusterLoadAssignment}.
+ * @param <U> the type of the primer
+ */
+abstract class ResourceHolderWithPrimer
+        <SELF extends ResourceHolderWithPrimer<SELF, T, U>, T, U extends ResourceHolder<?>>
+        implements ResourceHolder<T> {
+
+    abstract SELF withPrimer(@Nullable U primer);
+
+    @Nullable
+    abstract U primer();
```

Comment:
How about taking the resource type as is? It would make the type signature simpler.
```suggestion
abstract class ResourceHolderWithPrimer
        <SELF extends ResourceHolderWithPrimer<SELF, T, U>, T extends Message, U extends Message>
        implements ResourceHolder<T> {

    abstract SELF withPrimer(@Nullable ResourceHolder<U> primer);

    @Nullable
    abstract ResourceHolder<U> primer();
```

We also need to change the concrete types with parameterized ones.
```java
public final class RouteResourceHolder
        extends ResourceHolderWithPrimer<RouteResourceHolder, RouteConfiguration, Listener> {

    RouteResourceHolder(RouteConfiguration routeConfiguration, ResourceHolder<Listener> primer) {
        this.routeConfiguration = routeConfiguration;
        this.primer = primer;
    }
    ...

    @Override
    RouteResourceHolder withPrimer(@Nullable ResourceHolder<Listener> primer) {
        if (primer == null) {
            return this;
        }
        return new RouteResourceHolder(routeConfiguration, primer);
    }

    @Override
    @Nullable
    ResourceHolder<Listener> primer() {
        return primer;
    }
    
```